### DISCUSSION TITLE
Manageable document versions

### DISCUSSION THREAD

**Discussion by tauremacar** [umbraco/Umbraco-CMS#10514] [Features and ideas]
Umbraco have a sort of document versioning, but actually it's more like a change log and not very easy to manage the way we want.
We would like to see a more manageable kind of document versions, typically for content that are connected to a time span or have some kind of sequence. For us it's about content related to public laws and regulations. But it can also be user manuals/documentation, like on our.umbraco.com, or other content describing stuff that changes over time. The point is that the versions, or validity of the versions, are changing at a spesific point of time, like January 1st or on a release date. And that former/upcoming versions still may be valueable for users.

A feature like this should include things like this:
1. Possibility to work on multiple versions of the same document, each valid for a (not overlapping) period of time.
2. Possibility for end user to select both current and previous/upcoming versions (if published!). Both by version number and by date. URL parameter/postfix?
3. Current version must be default, both when serving it to end users and when editing in the backoffice.
4. Build in functionality must ensure that start/end dates and publish status is correct for all versions (like not overlapping dates).
5. Possibility to publish two or more versions, but only the one valid for current timespot must show up. At least as default.
6. The feature could be off by default, or not complicate things if you don't want to use it. 

I think it's several ways to customize this in a project, and it doesn't have to be super hard. But it's always a challenge to create a good user experience without connecting too tight to Umbraco, which can mess up during upcoming upgrades. So it would be very nice to have this feature well implemented within Umbraco itself.

I know this has been asked for some times over the years, like in the posts below (links), but as far as I know there is still no plans for this in upcoming versions. I think it should be, as this is a quite common feature in CMS systems.

Related posts:
https://our.umbraco.com/forum/umbraco-7/using-umbraco-7/73097-versioning-roadmap
https://our.umbraco.com/forum/using-umbraco-and-getting-started/104323-versioning-content

**Comment by nul800sebastiaan** [comment]
Thank you, this is something we're still thinking about and we agree we need to make this happen at some point. Currently that is not on the roadmap though and we have no timeframe to announce in which we can work on this.

**Reply by tauremacar** [reply]
Thanks, @nul800sebastiaan. It's good to see that you agree on this. I understand that there's a lot of things to do, especially these days. But this would make Umbraco an even greater CMS, both for developers, users and the market.

Anyway, we're about to convert from a certain Danish enterprice CMS (which have this feature) to Umbraco, so we need to handle this in a way. In version 9. We will work on this during the autumn, so maybe you will consider it. ;-) And maybe we can contribute in some way.

**Reply by nul800sebastiaan** [reply]
Contributions would be great but we need to give you guidance first since we don't want to waste your time implementing this and then it's not how we think it should be implemented. Speccing this out is also something we currently don't have time for though.

The alternative I can think of is to implement your own way. As I am reading this again I realize that you have very specific use cases here that I doubt we'll be covering. I am more thinking about 2 features we would really like:

1. Locking the content item if someone else is working on it so you don't overwrite each other's changes (with an "override, I know what I'm doing" button possibly)
2. Being able to edit content while a version of the content has been scheduled for publishing, where that scheduled version is absolutely locked for editing unless someone overrides that. Currently you can just keep editing until the last minute and the last saved version will be published, which is sometimes problematic

For your use cases it might be good to check out Plumber and think about how the features you mention are really used (in other words how you could bend the usage if needed.. 🙂).

Plumber overview: https://umbraco.com/apps/plumber/
Plumber package: https://our.umbraco.com/packages/backoffice-extensions/plumber-workflow-for-umbraco/

Maybe the author, @nathanwoulfe, could give you some tips on your desired workflows as well.

**Reply by tauremacar** [reply]
Thanks you for your info and time on this! I don't want to use much of your time when it's not on the agenda, though.
And by contributing I didn't necessarily mean coding, but maybe just ideas and other input when/if wanted.

I'm aware of the 'lack of lock' issue, but I don't think that will be a problem for our use. That one is also a bit scary for a first PR. :-)

About the versioning, I think we are talking about two different kind of solutions. Our thoughts is not to extend or interact with the current version system in Umbraco, but rather build something on top of it. Or on top of documents itself.
Like having a content type as a parent for other documents, which then will be the different versions. Each version will then still have their own versions/change log and other build-in stuff. The parent can act like a proxy for its document children, and have the responsibility for handling and serving the right version(s). Each version is technically a separate Umbraco document, but will be wrapped behind the parent. Creating a new version will basically be to copy one of the existing ones, and update some properties.

So far this is just thoughts, but our upcoming PoC will proof (or disproof) the concept. :-)
